src/probnum/utils/linalg/_inner_product.py

```diff
@@ -1,65 +1,76 @@
 """Functions defining useful inner products."""
+from __future__ import annotations
 
-from typing import Optional, Union
+from typing import TYPE_CHECKING, Optional, Union
 
 import numpy as np
 
-from probnum import linops
+if TYPE_CHECKING:
+    from probnum import linops
 
 
-def euclidean_inprod(
+def inner_product(
     v: np.ndarray,
     w: np.ndarray,
     A: Optional[Union[np.ndarray, linops.LinearOperator]] = None,
 ) -> np.ndarray:
-    r"""(Modified) Euclidean inner product :math:`\langle v, w \rangle_A := v^T A w`.
+    r"""Inner product :math:`\langle v, w \rangle_A := v^T A w`.
+
+    For arrays the function computes the inner product over the last axes of the
+    two arrays ``v`` and ``w``.
 
     Parameters
     ----------
     v
-        First vector.
+        First array.
     w
-        Second vector.
+        Second array.
     A
         Symmetric positive (semi-)definite matrix defining the geometry.
 
     Returns
     -------
-    inprod
-        Inner product.
+    inprod :
+        *shape=(\*v.shape[:-1], \*w.shape[:-1])* -- Inner product of ``v`` and ``w``. If they are both 1-D arrays then a scalar is returned; otherwise an array is returned.
```

Comment:
Personally, I find the `np.dot`-style broadcasting extremely difficult to work with, since it practically forces you to compute all pairwise inner products. Besides, it is incompatible with the 'stack of matrices'-style broadcasting used in `__matmul__` (thus also in `pn.linops`) and in `np.linalg`. So I would propose implementing these functions here to realize the einsum expression `"...i,...i"`, even though this deviates from `np.inner`. Note that this style of broadcasting is actually more powerful than the `np.dot`-style broadcasting, since
```python
a = ... # shape: (L, M, N)
b = ... # shape: (J, K, N)
np.inner(a, b) == np.einsum("...i,...i", a[:, :, None, None, :], b[None, None, :, :, :]) # shape: (L, M, J, K)
```
Also, this way the `inner_product` implementation would compatible with its `induced_norm` broadcasting-wise.
Sorry for the strong opinion, I just really hate the broadcasting rules in `np.inner`/`np.dot`/`np.tensordot` 😀